src/protocol/attribution/aggregate_credit.rs

```diff
@@ -160,6 +163,160 @@ where
+    let m_ctx = malicious_validator.context();
+
+    let one = m_ctx.share_known_value(F::ONE);
+
+    let mut stop_bits = repeat(one.clone()).take(num_rows).collect::<Vec<_>>();
+
+    let sorted_input = m_ctx.upgrade(sorted_input).await?;
+    let mut credits = sorted_input
+        .iter()
+        .map(|x| x.credit.clone())
+        .collect::<Vec<_>>();
+
+    for (depth, step_size) in std::iter::successors(Some(1_usize), |prev| prev.checked_mul(2))
+        .take_while(|&v| v < num_rows)
+        .enumerate()
+    {
+        let end = num_rows - step_size;
+        let c = m_ctx
+            .narrow(&InteractionPatternStep::from(depth))
+            .set_total_records(end);
+        let mut futures = Vec::with_capacity(end);
+
+        for i in 0..end {
+            let c = c.clone();
+            let record_id = RecordId::from(i);
+            let sibling_helper_bit = &sorted_input[i + step_size].helper_bit;
+            let current_stop_bit = &stop_bits[i];
+            let sibling_stop_bit = &stop_bits[i + step_size];
+            let sibling_credit = &credits[i + step_size];
+            futures.push(async move {
+                let b = compute_b_bit(
+                    c.narrow(&Step::ComputeBBit),
+                    record_id,
+                    current_stop_bit,
+                    sibling_helper_bit,
+                    depth == 0,
+                )
+                .await?;
+
+                try_join(
+                    c.narrow(&Step::AggregateCreditBTimesSuccessorCredit)
+                        .multiply(record_id, &b, sibling_credit),
+                    compute_stop_bit(
+                        c.narrow(&Step::ComputeStopBit),
+                        record_id,
+                        &b,
+                        sibling_stop_bit,
+                        depth == 0,
+                    ),
+                )
+                .await
+            });
+        }
+
+        let results = try_join_all(futures).await?;
+
+        results
+            .into_iter()
+            .enumerate()
+            .for_each(|(i, (credit, stop_bit))| {
+                credits[i] += &credit;
+                stop_bits[i] = stop_bit;
+            });
+    }
+
+    // Prepare the sidecar for sorting
+    let aggregated_credits = sorted_input
+        .iter()
+        .enumerate()
+        .map(|(i, x)| MCCappedCreditsWithAggregationBit {
+            helper_bit: x.helper_bit.clone(),
+            aggregation_bit: x.aggregation_bit.clone(),
+            breakdown_key: x.breakdown_key.clone(),
+            credit: credits[i].clone(),
+            _marker: PhantomData::default(),
+        })
+        .collect::<Vec<_>>();
+
+    let aggregated_credits = malicious_validator.validate(aggregated_credits).await?;
+    //
+    // 4. Sort by `aggregation_bit`
+    //
+    let sorted_output = sort_by_aggregation_bit(
+        sh_ctx.narrow(&Step::SortByAttributionBit),
+        aggregated_credits,
+    )
+    .await?;
```

Comment:
Is this a malicious sort?

Reply:
No, https://github.com/private-attribution/ipa/pull/457 should take care of that. Thanks for spotting this